controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -92,14 +134,335 @@ func (r *StorageClassClaimReconciler) Reconcile(ctx context.Context, request rec
+				r.log.Info("Waiting for CephFilesystem to be Ready. Skip reconciling StorageClass", "CephFilesystem", klog.KRef(storageClass.Parameters["fsName"], storageClass.Parameters["clusterID"]), "StorageClass", klog.KRef("", storageClass.Name))
+				skippedStorageClass = append(skippedStorageClass, storageClass.Name)
+				continue
+			}
+		}
+
+		// Any existing (or newly created) StorageClass will have the same name as the consumer-created StorageClassClaim, which will also own it.
+		storageClass.Name = r.storageClassClaim.Name
+		storageClass.OwnerReferences = []metav1.OwnerReference{storageClassClaimOwnerReference}
+		existing := &storagev1.StorageClass{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: r.storageClassClaim.Name,
+			},
+		}
+		err := r.get(existing)
+		if errors.IsNotFound(err) {
+
+			// Since the StorageClass is not found, we will create a new one.
+			r.log.Info("Creating StorageClass.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
+			err = r.Client.Create(r.ctx, storageClass)
+			if err != nil {
+				return err
+			}
+		} else if err != nil {
+			return err
+		} else {
+			if storageclassConfiguration.reconcileStrategy == "init" {
+				continue
+			}
+			if existing.DeletionTimestamp != nil {
+				return fmt.Errorf("failed to restore StorageClass  %s because it is marked for deletion", existing.Name)
+			}
+			if !reflect.DeepEqual(storageClass.Parameters, existing.Parameters) {
+
+				// Since we have to update the existing StorageClass, so we will delete the existing storageclass and create a new one.
+				r.log.Info("StorageClass needs to be updated, deleting it.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
+				err = r.Client.Delete(r.ctx, existing)
+				if err != nil {
+					r.log.Error(err, "Failed to delete StorageClass.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
+					return err
+				}
+				r.log.Info("Creating StorageClass.", "StorageClass", klog.KRef(storageClass.Namespace, storageClass.Name))
+				err = r.Client.Create(r.ctx, storageClass)
+				if err != nil {
+					r.log.Info("Failed to create StorageClass.", "StorageClass", klog.KRef(storageClass.Namespace, storageClass.Name))
+					return err
+				}
+			}
+		}
+	}
+	if len(skippedStorageClass) > 0 {
+		return fmt.Errorf("some StorageClasses [%s] were skipped while waiting for pre-requisites to be met", strings.Join(skippedStorageClass, ","))
+	}
+	return nil
+}
+
+func (r *StorageClassClaimReconciler) newCephBlockPoolStorageClassConfiguration(storageClassConfiguration StorageClassConfiguration) StorageClassConfiguration {
+	persistentVolumeReclaimDelete := corev1.PersistentVolumeReclaimDelete
+	allowVolumeExpansion := true
+	managementSpec := r.storageCluster.Spec.ManagedResources.CephBlockPools
+	storageClass := &storagev1.StorageClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("%s-ceph-rbd", r.storageCluster.Name),
```

Comment:
This name is just for cluster default storage classes, which does not make sense as in most cases this will be wrong. 
Please pass all the needed configurations as arguments and utilize them.